### ISSUE TITLE
Possible race when creating `/run/systemd/unit-root/run/systemd/incoming`(?)

### ISSUE THREAD

**Issue by mrc0mmand** [systemd/systemd#21907]
**systemd version the issue has been seen with**
 > latest main

**Used distribution**
 > CentOS 8

**Linux kernel version used** (`uname -a`)

 > `Linux version 4.18.0-348.2.1.el8_5.x86_64 (mockbuild@kbuilder.bsys.centos.org) (gcc version 8.5.0 20210514 (Red Hat 8.5.0-4) (GCC)) #1 SMP Tue Nov 16 14:42:35 UTC 2021`

When triaging CentOS CI fails I noticed a (so far) one-time fail in TEST-49 which might be an actual issue (albeit I'm not sure if in systemd, kernel, or somewhere else):

```
$ journalctl --file system.journal --no-hostname -o short-monotonic _PID=33
...
[  957.112629] systemd[33]: Successfully mounted /run/systemd/inaccessible/dir to /run/systemd/unit-root/run/inaccessible
[  957.112815] systemd[33]: Applying namespace mount on /run/systemd/unit-root/run/systemd/incoming
[  957.112922] systemd[33]: Followed source symlinks /run/systemd/propagate/testsuite-49-namespaced.service → /run/systemd/propagate/testsuite-49-namespaced.service.
[  957.113129] systemd[33]: Bind-mounting /run/systemd/propagate/testsuite-49-namespaced.service on /run/systemd/unit-root/run/systemd/incoming (MS_BIND "")...
[  957.113246] systemd[33]: Failed to mount /run/systemd/propagate/testsuite-49-namespaced.service (type n/a) on /run/systemd/unit-root/run/systemd/incoming (MS_BIND ""): No such file or directory
[  957.113358] systemd[33]: Failed to create destination mount point node '/run/systemd/unit-root/run/systemd/incoming': File exists
[  957.113478] systemd[33]: Failed to mount /run/systemd/propagate/testsuite-49-namespaced.service to /run/systemd/unit-root/run/systemd/incoming: No such file or directory
[  957.113647] systemd[33]: testsuite-49-namespaced.service: Failed to set up mount namespacing: /run/systemd/unit-root/run/systemd/incoming: No such file or directory
[  957.113766] systemd[33]: testsuite-49-namespaced.service: Failed at step NAMESPACE spawning grep: No such file or directory
```

Full journal: [system.journal.tar.gz](https://github.com/systemd/systemd/files/7777309/system.journal.tar.gz)

Log from a successful run, just for comparison:
```
...
[    1.889841] systemd[578]: Successfully mounted /run/systemd/inaccessible/dir to /run/systemd/unit-root/run/inaccessible
[    1.889918] systemd[578]: Applying namespace mount on /run/systemd/unit-root/run/systemd/incoming
[    1.890000] systemd[578]: Followed source symlinks /run/systemd/propagate/testsuite-49-namespaced.service → /run/systemd/propagate/testsuite-49-namespaced.service.
[    1.890094] systemd[578]: Bind-mounting /run/systemd/propagate/testsuite-49-namespaced.service on /run/systemd/unit-root/run/systemd/incoming (MS_BIND "")...
[    1.890188] systemd[578]: Successfully mounted /run/systemd/propagate/testsuite-49-namespaced.service to /run/systemd/unit-root/run/systemd/incoming
[    1.890276] systemd[578]: Applying namespace mount on /run/systemd/unit-root/sys
[    1.890446] systemd[578]: Applying namespace mount on /run/systemd/unit-root/tmp
...
```

I'm posting this mainly just to have some information about the issue in case it happens again in the future.

**Comment by poettering** [comment]
Hmm, that dir is inside a mount namespace that noone but the nspawn instance can see, i.e. its /run is private to the container, and freshly mounted, hence noone else should be able to create /run/systemd/incoming below it. Weird stuff.

**Comment by sta-c0000** [comment]
I don't know if this is related, but I've encountered an issue with two services starting nearly at the same time.
I'm only seeing this on Debian unstable (sid), where things can be indeed be unreliable, but I though maybe I should mention it here in case.  Apologies if it's not pertinent, or due to something else.

Test system is a slow i686 PC running Debian sid (unstable) with:
```
systemd 250 (250.3-2)
+PAM +AUDIT +SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS -OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY -P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK -XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified

Linux NB 5.16.0-2-686 #1 SMP PREEMPT Debian 5.16.10-1 (2022-02-18) i686 GNU/Linux
```

The two services are both pre-network, very similar oneshots: [nftables.service](https://sources.debian.org/src/nftables/1.0.2-1/debian/nftables.service/) and [so_mark_giduid.service](https://github.com/sta-c0000/galaf/blob/main/examples/so_mark_giduid.service)

Sometimes on boot `nftables.service` and `so_mark_giduid.service` **both** start without any issues (OK):
```
systemd[1]: Starting nftables...
systemd[1]: Starting Load and attach so_mark_giduid eBPF...
systemd[1]: File System Check on Root Device was skipped because of a failed condition check (ConditionPathExists=!/run/initramfs/fsck-root).
systemd[1]: Starting Journal Service...
systemd[1]: Starting Load Kernel Modules...
systemd[1]: Starting Remount Root and Kernel File Systems...
systemd[1]: Repartition Root Disk was skipped because all trigger condition checks failed.
systemd[1]: Starting Coldplug All udev Devices...
systemd[1]: Finished Create List of Static Device Nodes.
systemd[1]: modprobe@configfs.service: Deactivated successfully.
systemd[1]: Finished Load Kernel Module configfs.
systemd[1]: Mounting Kernel Configuration File System...
kernel: fuse: init (API version 7.35)
systemd[1]: modprobe@fuse.service: Deactivated successfully.
systemd[1]: Finished Load Kernel Module fuse.
systemd[1]: Mounting FUSE Control File System...
systemd[1]: Finished Load Kernel Modules.
systemd[1]: Starting Apply Kernel Variables...
kernel: ACPI: bus type drm_connector registered
systemd[1]: modprobe@drm.service: Deactivated successfully.
systemd[1]: Finished Load Kernel Module drm.
systemd-journald[218]: Journal started
systemd-journald[218]: Runtime Journal (/run/log/journal/4a0e1e6618e84ae1875cae767626ab1e) is 1.2M, max 9.8M, 8.6M free.
systemd[1]: Started Journal Service.
systemd[1]: Finished Load and attach so_mark_giduid eBPF.
[... omitted block; nftables takes couple seconds to finish; is normal ...]
systemd[1]: Finished nftables.
systemd[1]: Reached target Preparation for Network.
```
Without changing anything, just **lots of** reboots... (I have not been able to reproduce after boot.)
Sometimes on boot `nftables.service` fails with a `226/NAMESPACE` error, and `so_mark_giduid.service` starts OK:
```
systemd[1]: Starting nftables...
systemd[1]: Starting Load and attach so_mark_giduid eBPF...
systemd[1]: File System Check on Root Device was skipped because of a failed condition check (ConditionPathExists=!/run/initramfs/fsck-root).
systemd[1]: Starting Journal Service...
systemd[214]: Failed to create destination mount point node '/run/systemd/unit-root/run/systemd/incoming': File exists
systemd[214]: Failed to mount /run/systemd/propagate/nftables.service to /run/systemd/unit-root/run/systemd/incoming: No such file or directory
systemd[214]: nftables.service: Failed to set up mount namespacing: /run/systemd/unit-root/run/systemd/incoming: No such file or directory
systemd[1]: Starting Load Kernel Modules...
systemd[1]: Starting Remount Root and Kernel File Systems...
systemd[1]: Repartition Root Disk was skipped because all trigger condition checks failed.
systemd[1]: Starting Coldplug All udev Devices...
systemd[1]: Finished Create List of Static Device Nodes.
systemd[1]: modprobe@configfs.service: Deactivated successfully.
systemd[1]: Finished Load Kernel Module configfs.
systemd[1]: nftables.service: Main process exited, code=exited, status=226/NAMESPACE
systemd[1]: nftables.service: Failed with result 'exit-code'.
systemd[1]: Failed to start nftables.
```
Sometimes on boot `nftables.service` starts OK, and `so_mark_giduid.service` fails with a `226/NAMESPACE` error:
```
systemd[1]: Starting nftables...
systemd[1]: Starting Load and attach so_mark_giduid eBPF...
systemd[1]: File System Check on Root Device was skipped because of a failed condition check (ConditionPathExists=!/run/initramfs/fsck-root).
systemd[1]: Starting Journal Service...
systemd[215]: Failed to create destination mount point node '/run/systemd/unit-root/run/systemd/incoming': File exists
systemd[215]: Failed to mount /run/systemd/propagate/so_mark_giduid.service to /run/systemd/unit-root/run/systemd/incoming: No such file or directory
systemd[215]: so_mark_giduid.service: Failed to set up mount namespacing: /run/systemd/unit-root/run/systemd/incoming: No such file or directory
systemd[1]: Starting Load Kernel Modules...
systemd[1]: Starting Remount Root and Kernel File Systems...
systemd[1]: Repartition Root Disk was skipped because all trigger condition checks failed.
systemd[1]: Starting Coldplug All udev Devices...
systemd[1]: Finished Create List of Static Device Nodes.
systemd[1]: modprobe@configfs.service: Deactivated successfully.
systemd[1]: Finished Load Kernel Module configfs.
systemd[1]: so_mark_giduid.service: Main process exited, code=exited, status=226/NAMESPACE
systemd[1]: so_mark_giduid.service: Failed with result 'exit-code'.
systemd[1]: Failed to start Load and attach so_mark_giduid eBPF.
```
Simply commenting out `ProtectHome=true` from `so_mark_giduid.service` appears to eliminate the problem?
```
#ProtectHome=true
```
And then they always (_at least so far_) both load at boot without issues?
Otherwise **sometimes** they could step on each other's feet?
Thank you.

**Comment by yuwata** [comment]
Should be already fixed by 9d6d4c305ab8d65aab7f546450d7331f760b7259. Closing.
